## Distributions Induced by a Policy

 - an infinite-horizon MDP M = <S, A, R, T, $\gamma$>
     - S : states
     - A : actions
     - R : rewards
     - T : transitions
     - $\gamma$ : discount factor
 - stochastic policies of the form $\pi: S \rightarrow \Delta(A)$  
 - For a finite set X, $\Delta(X)$ refers to the set of categorical distributions with support on X or, equivalently, the $\Delta^{|X|-1}$ probability simplex.}.     
 - Additionally, we'll assume that M has a single, fixed starting state $s_0 \in S$ for simplicity.

### 1-a

Suppose we have a single Markov Decision Process (MDP) and two policies for that MDP, π1 and π2.
Naturally, we are often interested in the performance of policies obtained in the MDP, quantified by V^{π1} and V^{π2} , respectively.
If the reward function and transition dynamics of the underlying MDP are known to us, we can use standard methods for policy evaluation.
There are many scenarios, however, where the underlying MDP model is not known and we must try to infer something about the performance of policy π2 solely based on data obtained through executing policy π1 within the environment.
In this problem, we will explore a classic result for quantifying the gap in performance between two policies that only requires access to data sampled from one of the policies.

Consider 
1. an infinite-horizon MDP M = ⟨S, A, R, P, γ⟩.
	1-1. S is a set of Markov states s in S
	1-2. A is a set of actions a in A
	1-3. R is a reward function
	1-4. P is the dynamics/transition model for each action, that specifies P(st+1 =s′|st =s,at =a)
	1-5. γ is the discount factor
2. stochastic policies of the form π : S → ∆(A)
	For a finite set X, the notation ∆(X) refers to the set of categorical distributions with support on X or, equivalently, the ∆^{|X|−1} probability simplex.

Specifically, π(a|s) refers to the probability of taking action a in state s, and \sum_{a} π(a|s) = 1, ∀s.
For simplicity, we’ll assume that this decision process has a single, fixed starting state s0 in S.

Consider a fixed stochastic policy and imagine running several rollouts of this policy within the environment.
Naturally, depending on the stochasticity of the MDP M and the policy itself, some trajectories are more likely than others.
Question:
Write down an expression for ρ^{π}(τ), the probability of sampling a trajectory τ = (s0, a0, s1, a1, . . .) from running policy π in the MDP M.
To put this distribution in context, recall that $V^{π} (s0) = \mathbb{E}_{τ~ρ^{π}} [ \sum_{t=0}^{t=\infty} γ^t * R(s_t,a_t) | s0 ]$

### Response 1-(a)

The likelihood of sampling a trajectory $\tau = (s_0,a_0,s_1,a_1,\ldots)$ by running a stochastic policy $\pi$ in the MDP $M$ is given by the product of the probabilities of each state-action pair along the trajectory:

$\rho^{\pi}(\tau) = P(s_0) \prod_{t=0}^{\infty} \pi(a_t | s_t) P(s_{t+1} | s_t, a_t) $

The product is taken over time, representing the sequential nature of the trajectory.

Breaking it down:
 - $P(s_0)$ is the probability of the starting state $s_0$ under the policy $\pi$.
 - $ \pi(a_t | s_t)$ is the probability of taking action $a_t$ in state $s_t$ under the policy $\pi$.
 - $P(s_{t+1} | s_t, a_t)$ is the probability of transitioning to state $s_{t+1}$ given that action $a_t$ was taken in state $s_t$ under the dynamics of the MDP.

These probabilities are determined by the stochastic policy $\pi$ and the transition dynamics $T$ of the MDP. The expression represents the likelihood of the entire trajectory $\tau$ under the policy $\pi$.

The expectation in the value function $V^{π} (s0) = \mathbb{E}_{τ~ρ^{π}} [ \sum_{t=0}^{t=\infty} γ^t * R(s_t,a_t) | s0 ]$ involves summing over all possible trajectories weighted by their likelihoods, which is why the distribution of trajectories $\rho^\pi$ is needed for evaluation.

What is $p^{\pi}(s_t = s)$, where $p^{\pi}(s_t = s)$ denotes the probability of being in state s at timestep t while following policy π? (Provide an equation)

$p^{\pi}(s_t = s) = \sum_{\tau} \rho^{\pi}(\tau) \cdot \mathbb{I}(s_t = s)$


$\rho^{\pi}(\tau)$ is the probability of sampling the trajectory $\tau$ under policy $\pi$.  
$\mathbb{I}(s_t = s)$ is the indicator function, which is 1 if $s_t = s$ and 0 otherwise.

### 1-(b)

Just as ρ^π captures the distribution over trajectories induced by policy π, we can also examine the distribution over states induced by policy π.
In particular, define the "discounted, stationary state distribution" of a policy π as, $$d^{π}(s) = (1-γ) \sum\limits_{t=0}^\infty γ^t p(s_t = s),$$ where $p(s_t = s)$ denotes the probability of being in state $s$ at timestep t while following policy π.
Your answer to the previous part should help you reason about how you might compute this value.  

The value function of a policy π can be expressed using this distribution d^{π}(s,a) = d^{π}(s) * π(a|s) over states and actions, which will shortly be quite useful.

Consider an arbitrary function $f: S \times A \rightarrow R$.   
Prove the following identity:   
$$\mathbb{E}_{τ \sim ρ^π} \left [\sum\limits_{t=0}^\infty γ^t f(s_t,a_t)\right] = 1/(1-γ)* \mathbb{E}_{s \sim d^π}[ \mathbb{E}_{ a ~ π(s) }[f(s,a)] ]$$ 

Hint 0: You may find it helpful to first consider how things work out for f(s, a) = 1, for all (s, a) in S × A.  
Hint 1: Using your answer from part (a), try to understand the following identity $\mathbb{E}_{\tau \sim \rho^{\pi}}[1[s_{t}=s]] = p(s_{t} = s)$}.  
Hint 2: Recall the linearity property of the expectation operator.

### Response 1-(b)

First, let's consider the case when $ f(s, a) = 1 $ for all $ (s, a) \in S \times A$.
#### Case 1: $f(s, a) = 1$

The left-hand side (LHS) of the identity is:
$
\mathbb{E}_{\tau \sim \rho^{\pi}} \left [\sum_{t=0}^{\infty} \gamma^t \cdot 1 \right]
$

Using Hint 1, we know that $\mathbb{E}_{\tau \sim \rho^{\pi}}[1[s_{t}=s]] = p(s_{t} = s)$.  
Substituting this in, we get:
$\mathbb{E}_{\tau \sim \rho^{\pi}} \left [\sum_{t=0}^{\infty} \gamma^t \cdot 1[s_{t}=s] \right] = \sum_{t=0}^{\infty} \gamma^t \cdot p(s_{t} = s)$

Now, let's look at the right-hand side (RHS) of the identity:
$\frac{1}{1-\gamma} \cdot \mathbb{E}_{s \sim d^{\pi}} [1]$

The inner expectation $\mathbb{E}_{a \sim \pi(s)}[1]$ is just the sum of probabilities of all actions in state s according to policy $\pi$, which is equal to 1.   
Therefore, the RHS simplifies to:
$\frac{1}{1-\gamma} \cdot \mathbb{E}_{s \sim d^{\pi}} [1]$

Now, the term $\mathbb{E}_{s \sim d^{\pi}} [1]$ represents the sum of discounted probabilities of being in state s under policy $\pi$, which is equivalent to the expression we obtained for the LHS.   
Therefore, in the case when $f(s, a) = 1$, the identity holds.



#### Case 2: Arbitrary $f: S \times A \rightarrow \mathbb{R}$

Now, let's extend the proof to an arbitrary function \(f: S \times A \rightarrow \mathbb{R}\).

The LHS of the identity is:
$
\mathbb{E}_{\tau \sim \rho^{\pi}} \left [\sum_{t=0}^{\infty} \gamma^t \cdot f(s_t, a_t) \right]
$

Using linearity of the expectation operator, we can express this as:
$
\sum_{t=0}^{\infty} \gamma^t \cdot \mathbb{E}_{\tau \sim \rho^{\pi}}[f(s_t, a_t)]
$

Now, applying the result from Case 1 for the expectation $\mathbb{E}_{\tau \sim \rho^{\pi}}[1[s_{t}=s]] = p(s_{t} = s)$, we have:
$
\sum_{t=0}^{\infty} \gamma^t \cdot \sum_{s \in S} p(s_t = s) \cdot \mathbb{E}_{a \sim \pi(s)}[f(s, a)]
$

We konw the definition of the discounted, stationary state distribution is $$d^{π}(s) = (1-γ) \sum\limits_{t=0}^\infty γ^t p(s_t = s),$$  
By this definision, we rearrange the expression of LFS to
$
\frac{1}{1-\gamma} \sum_{s \in S} d^{\pi}(s) \cdot \mathbb{E}_{a \sim \pi(s)}[f(s, a)]
$

Finally, the RHS of the identity is:
$
\frac{1}{1-\gamma} \cdot \mathbb{E}_{s \sim d^{\pi}} \left[ \mathbb{E}_{a \sim \pi(s)}[f(s, a)] \right]
$

Comparing the LHS and RHS expressions, we see that the identity holds for arbitrary $f: S \times A \rightarrow \mathbb{R}$. QED.

### 1-(c)

For any policy $\pi$, we define the following function $$A^\pi(s,a) = Q^\pi(s,a) - V^\pi(s)$$.  
$A^\pi(s,a)$ is known as the advantage function and shows up in a lot of policy gradient based RL algorithms.  
Intuitively, it is the additional benefit one gets from first following action a and then following π, instead of always following π.


Prove the following statement holds for all policies $\pi,\pi'$: 

$$V^\pi(s_0) - V^{\pi'}(s_0) = \frac{1}{(1-\gamma)} \mathbb{E}_{s \sim d^\pi}\left[\mathbb{E}_{a \sim \pi(s)}\left[A^{\pi'}(s,a)\right]\right]$$


Hint 1: Recall the "tower property of expectation" which says that E[X] = E[E[X|Y]]

Hint 2: $$\mathbb{E}_{τ \sim ρ^π} \left [\sum\limits_{t=0}^\infty γ^t f(s_t,a_t)\right] = 1/(1-γ)* \mathbb{E}_{s \sim d^π}[ \mathbb{E}_{ a \sim π(s) }[f(s,a)] ]$$ 

Note: We have provided you with scaffolding for your derivation. In the provided scaffold we have substituted in our expression for $V^{\pi}(s_{0})$ from part (a). In addition, we have reexpressed $V^{\pi'}(s_{0})$ as a telescoping sum.  

$V^\pi(s_0) - V^{\pi'}(s_0) = \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t \mathcal{R}(s_t,a_t)\right] - V^{\pi'}(s_0) $  
$= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( \mathcal{R}(s_t,a_t) + V^{\pi'}(s_t) - V^{\pi'}(s_t)\right)\right] - V^{\pi'}(s_0) $  
$= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( \mathcal{R}(s_t,a_t) + \gamma V^{\pi'}(s_{t+1}) - V^{\pi'}(s_t)\right)\right] $

### Response 1-(c)

\begin{align*}
V^\pi(s_0) - V^{\pi'}(s_0) &= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t \mathcal{R}(s_t,a_t)\right] - V^{\pi'}(s_0) \\
&= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( \mathcal{R}(s_t,a_t) + V^{\pi'}(s_t) - V^{\pi'}(s_t)\right)\right] - V^{\pi'}(s_0) \\
&= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( \mathcal{R}(s_t,a_t) + \gamma V^{\pi'}(s_{t+1}) - V^{\pi'}(s_t)\right)\right] \\
&= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( Q^{\pi'}(s_t,a_t) - V^{\pi'}(s_t)\right)\right] \\
&= \mathbb{E}_{\tau \sim \rho^\pi}\left[\sum\limits_{t=0}^\infty \gamma^t\left( A^\pi(s_t,a_t)\right)\right] \\
&= 1/(1-γ)* \mathbb{E}_{s \sim d^π}[ \mathbb{E}_{ a \sim π(s) }[A^\pi(s,a)] ]
\end{align*}


### 2-(a) [4points]

1. objective function : define $V(s,n)$ as the maximum sum of rewards starting from state s in a single trajectory/episode of length n.  
2. recurrence relation : $\max_{a} ( r(s,a) + V( s', n-1 ) )$
    - taking an action **a** at state **s** will result in transitioning to state **s'** with an immediate reward **r(s,a)**  
3. boundary conditions : $V(s,0) = 0, \forall s$  
4. represent output as the objective function : $V(0,5)$  

In [7]:
import numpy as np

class Problem2:
    def __init__(self):
        self.transition = np.array([
            [[0, 0.2], [1, -0.1], [2, 0.0], [3, -0.3], [0, 0.2]],
            [[0, 0.2], [1, -0.1], [2, 0.0], [3, -0.3], [1, -0.1]],
            [[0, -2.0], [1, 1.0], [2, 0.0], [3, 3.0], [2, 0.0]],
            [[0, 0.2], [1, -0.1], [2, 0.0], [3, -0.3], [3, -0.3]]
        ])
        self.nextState = np.asarray( self.transition[:,:,0],dtype=int)
        self.immediateReward = np.asarray( self.transition[:,:,1],dtype=float)
        self.nA = 5 # number of actions
        self.nS = 4 # number of states
    
    """
    Input Arguments
        s0 : (int) : the starting state
        L : (int) : the length of the episode/trajectory.
    Returned Output:
        max_reward : float : the maximum sum of rewards starting from state s in a single trajectory/episode of length L.
        opt_actions : (list) of (int) : a sequence of states, representing the trajectory that renders the max_reward.
    """
    def DP ( self, s0, L ):
        valueTable = np.full( ( self.nS, L+1 ), float('-inf'))
        actionTable = np.zeros( ( self.nS, L+1 ), dtype=int )

        # boundary conditions
        if L == 0:
            return 0, []
        for state in range(self.nS):
            valueTable[state][0] = 0
        
        ### PART 1 TABLE FILLING ###
        # len_ : 1, 2, 3, ... L
        for len_ in range( 1, L+1 ):
            # state : 0, 1, 2, 3
            for state in range(self.nS):
                # action : 0, 1, 2, 3, 4
                for action in range(self.nA):
                    next_state = self.nextState[state][action]
                    immediate_reward = self.immediateReward[state][action]
                    if immediate_reward + valueTable[next_state][len_-1] > valueTable[state][len_]:
                        valueTable[state][len_] = immediate_reward + valueTable[next_state][len_-1]
                        actionTable[state][len_] = action
        
        ### PART 2 BACK-TRACKING ###
        optimal_actions = [0] * L
        state = s0

        for timestamp in range(L):
            opt_action = actionTable[state][L-timestamp]
            next_state = self.nextState[state][opt_action]
            optimal_actions[ timestamp ] = opt_action
            state = next_state

        return valueTable[s0][L], optimal_actions

    def get_traj( self, start_state, optimal_actions ):
        current_state = start_state
        trajectory = []

        for action_index in optimal_actions:
            next_state, reward = self.transition[current_state][action_index]
            trajectory.append({
                'state_t': current_state,
                'action_t': action_index,
                'reward_t': reward
            })
            current_state = int(next_state)

        return trajectory


In [8]:
num_steps = 5
starting_state = 0

p = Problem2()
max_sum_of_rewards, optimal_actions = p.DP(0,num_steps)
trajectory = p.get_traj(starting_state, optimal_actions)

print("\n - maximum sum of rewards starting from state " +str(starting_state)+ " in a single trajectory/episode of length " + str(num_steps) + " is " + str(max_sum_of_rewards) )
print(" - this value is attainable in a single trajectorythe by following the optimal sequence of actions:", optimal_actions)
print(" - the trajectory of following the optimal actions can be represented as st, at, Rt.")
for t, step in enumerate(trajectory):
    print(f"\tTimestamp {t} - State: {step['state_t']}, Action: {step['action_t']}, Reward: {step['reward_t']}")


 - maximum sum of rewards starting from state 0 in a single trajectory/episode of length 5 is 6.2
 - this value is attainable in a single trajectorythe by following the optimal sequence of actions: [0, 2, 3, 2, 3]
 - the trajectory of following the optimal actions can be represented as st, at, Rt.
	Timestamp 0 - State: 0, Action: 0, Reward: 0.2
	Timestamp 1 - State: 0, Action: 2, Reward: 0.0
	Timestamp 2 - State: 2, Action: 3, Reward: 3.0
	Timestamp 3 - State: 3, Action: 2, Reward: 0.0
	Timestamp 4 - State: 2, Action: 3, Reward: 3.0


The argument for why no other trajectory can achieve a greater cumulative reward is based on the design of the dynamic programming algorithm.
The algorithm explores all possible trajectories from the initial state with a fixed number of steps, ensuring that all potential paths are considered.
 - **the upper bound of the cumulated reward in a single trajectory**  
    - the maximum reward in a single step we can obtain is 3, by taking action "3" at state "2" then transitioning to state "3".
    - having transitioned to state "3", we need to transition back to state "2" from state "3" in order to obtain the maximum single-step reward 3 again.
    - under our scenario, we can get  at most $\lfloor {n_{step}/2} \rfloor$ = $\lfloor {5/2} \rfloor$ = 2 times of this max single-step reward.
    - by the above induction, the maximum reward we can get in 4 steps is 2 * $r_{max}$ = +6. ( --> state 2 --(+3)--> state 3 --> state 2 --(+3)--> state 3)
    - the max reward starting from state 0 is 0.2 
     ```
     np.max( self.immediateReard[0,:,:] )
     ```
    - Consequently, the overall upper bound of the cumulated reward in a single episode is 6 + 0.2 = 6.2.  
    - fact 1 : no trajectory can have a cumulated reward larger than 6.2.
    - fact 2: our optimal trajectory has a cumulated reward of 6.2.
    - conclusion : no other trajectory can achieve greater cumulative reward than the one we get, i.e. 6.2.
 - **Optimal Substructure**:  
    The dynamic programming approach considers the optimal substructure property, meaning that the optimal solution to the problem can be constructed from the optimal solutions of its subproblems. In this case, the optimal trajectory for a given state and remaining steps is built upon the optimal trajectories for its successor states with fewer steps.  
     - suppose the start state is $s_0$, and length of the episode is $L$.  
     - since a list of actions maps to a trajectory, we will discuss list of actions in the following discussion.  
     - suppose the optimal list of actions is P, and ${s_0,s_1,s_2,...s_{L}}$ is the sequence of states if we follow P.  
     - Let's consider an action $a_t$ on P for some t between 0 and $L-1$.  
     - What can we say for certain about cumulated reward from $s_0$ to $s_t$ and the cumulated reward from $s_{t+1}$ to $s_L$?  
     - For certain, we can say that the path from $s_0$ to $s_1$ must be the maximum cumulated reward of a trajectory of length $t-1$ from $s_0$ to $s_{t}$, and similarly the cumulated reward from $s_{t+1}$ to $s_{L}$ must be "a" possible path from $s_{t+1}$ to $s_{L}$ with max cumulated reward.  
     - Why?  We could argue by contradiction.  
     - Suppose that there was a path from $s_0$ to $s_t$ with larger cumulated reward.  
     - Then we could take that path, then follow the path from k to v and we would obtain a shorter path than P from u to v. This would contradict that u to v is a shortest path.
     is one and then consider any vertex k on the P somewhere between u and v.
     For certain, we can say that the accumulated from u to k must be a shortest possible path from u to k, and similarly the path from k to v must be a shortest possible path from k to v.
 - **Maximizing Cumulative Reward**:  
    At each step, the algorithm chooses the action that maximizes the cumulative reward. By doing this for every step in the trajectory, the algorithm is guaranteed to find the trajectory with the maximum cumulative reward among all possibilities.
 - **Backtracking and Memoization**:  
    The algorithm uses backtracking to explore different actions at each step and memoization to avoid redundant calculations for already explored subproblems. This ensures that each subproblem is only solved once, preventing unnecessary recomputation.
Given these factors, the algorithm is exhaustive in its exploration and selects the trajectory with the maximum cumulative reward at each step. Therefore, by the end of the exploration, the trajectories identified with the maximum sum of rewards are guaranteed to be the optimal trajectories, and no other trajectory can achieve a greater cumulative reward within the given constraints.

### 3-b. Tabular Q-Learning [5 points]

If the state and action spaces are sufficiently small, we can simply maintain a table containing the value of $Q(s,a)$, an estimate of $Q^*(s,a)$, for every $(s,a)$ pair.
In this tabular setting, given an experience sample $(s, a, r, s')$, the update rule is
\begin{align*}
Q(s,a) \leftarrow Q(s,a) + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}}Q(s',a') - Q(s,a)\right)
\end{align*}
where $\alpha > 0$ is the learning rate, $\gamma \in [0,1)$ the discount factor.

In addition, to formalizing our update rule for Q-learning in the tabular setting we must also consider strategies for exploration.  
In this question, we will be considering an $\epsilon$-greedy exploration strategy. This strategy means that each time we look to choose an action $A$, we will do so as follows,

\begin{align*}
A \leftarrow \begin{cases}
				argmax_{a \in \mathcal{A}} Q(s,a) \;\; \text{with probability $1-\epsilon$} \\
				a \in \mathcal{A}  \;\; \text{chosen uniformly at random with probability $\epsilon$}
			 \end{cases}
\end{align*}

In the following questions, you will need to implement both the update rule and $\epsilon$-greedy exploration strategy for Q-learning in the tabular setting.

We will now examine the issue of overestimation bias in Q-learning.   
The crux of the problem is that, since we take a max over actions, errors which cause Q to overestimate will tend to be amplified when computing the target value, while errors which cause Q to underestimate will tend to be suppressed.

Assume for simplicity that our Q function is an unbiased estimator of $Q^*$, meaning that $\mathbb{E}[Q(s,a)] = Q^*(s,a)$ for all states $s$ and actions $a$.  
Show that, even in this seemingly benign case, the estimator overestimates the real target in the following sense:
$$
\forall s, \quad \mathbb{E}\left[ \max_a Q(s,a) \right] \ge \max_a Q^*(s,a)
$$


Note: The expectation $\mathbb{E}[Q(s,a)]$ is over the randomness in $Q$ resulting from the stochasticity of the exploration process.

### Response 3-(b)
Since the **max** function is [convex](https://en.wikipedia.org/wiki/Convex_function), as per the [Jensen's Ineqality](https://en.wikipedia.org/wiki/Jensen%27s_inequality#:~:text=In%20mathematics%2C%20Jensen%27s%20inequality%2C%20named,integral%20of%20the%20convex%20function.) we have : 
$\mathbb{E}\left[\max_a Q(s,a)\right] >= \max_a \mathbb{E}\left[ Q(s,a)\right]$  

Since Q is an unbiased estimator of $Q^*$, we have : 
$\max_a \mathbb{E}\left[ Q(s,a)\right] = \max_a Q^*(s,a)$.  

Therefore, $\mathbb{E}\left[\max_a Q(s,a)\right] >= \max_a \mathbb{E}\left[ Q(s,a)\right] = \max_a Q^*(s,a)$.  

The expectation of the maximum over actions is greater than or equal to the maximum of the true $Q^*$ values for that state $s$.   
This shows that, even in the case where $Q$ is an unbiased estimator of $Q^*$, the estimator overestimates the real target.


### 5-(a) Linear Approximation

Suppose we represent the Q function as
$ Q_{\theta}(s, a) = \theta^\top \delta(s,a) $
 - $\theta \in \mathbb{R}^{\vert S \vert \vert A \vert \times 1 }$  
 - $\delta : S \times \mathcal{A} \rightarrow \mathbb{R}^{\vert S \vert \vert A \vert \times 1}$  

$
[\delta(s,a)]_{s'',a''} = 
\begin{cases} 
1 & \text{if } s'' = s \text{ and } a'' = a \\
0 & \text{otherwise}
\end{cases}
$


### Response 5-(a)-1
  1. Compute $\nabla_{\theta} Q_{\theta}(s, a)$


To apply the chain rule, we consider the composition of two functions:

1. The outer function $f(\theta) = \theta^\top \delta(s, a)$
2. The inner function $g(\theta) = \theta$ (which is a linear mapping)

The chain rule states that the derivative of the composition of two functions is the product of the derivative of the outer function and the derivative of the inner function. Mathematically, it is expressed as:

$ \nabla_{\theta} (f \circ g)(\theta) = (\nabla_{g} f)(g(\theta)) \cdot (\nabla_{\theta} g)(\theta) $

In our case:

$ \nabla_{\theta} Q_{\theta}(s, a) = (\nabla_{\theta} (\theta^\top \delta(s, a)) = (\nabla_{\theta} (\theta^\top))(g(\theta)) \cdot (\nabla_{\theta} g)(\theta) $

Now, $\nabla_{\theta} (\theta^\top)$ is the derivative of the transpose of $\theta$, which is simply the identity matrix I.

So, we have:

$ \nabla_{\theta} Q_{\theta}(s, a) = I \cdot \delta(s, a) = \delta(s, a) $

In summary, given the linear approximation $Q_{\theta}(s, a) = \theta^\top \delta(s, a)$ where $\delta(s, a)$ is a one-hot encoded vector, we have:

$\nabla_{\theta} Q_{\theta}(s, a) = \delta(s, a) $

### Response 5-(a)-2 Write the update rule for $\theta$.




Now, let's compare the tabular Q-learning update rule with the linear approximation update rule:

1. **Tabular Q-learning update rule:**
$Q(s,a) \leftarrow Q(s,a) + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}}Q(s',a') - Q(s,a)\right)$

2. **Linear approximation update rule:**
$\theta \leftarrow \theta + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} Q_{\theta}(s', a') - Q_{\theta}(s, a)\right) \nabla_{\theta}Q_{\theta}(s, a)$

Now, substitute the expressions for $Q_{\theta}(s, a)$ and $\nabla_{\theta} Q_{\theta}(s, a)$ into the linear approximation update rule:

$\theta \leftarrow \theta + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} (\theta^\top \delta(s', a')) - \theta^\top \delta(s, a)\right) \delta(s, a)$

### Response 5-(a)-3
  3. Argue that equation for the tabular q-learning update rule we saw before:  
  $Q(s,a) \leftarrow Q(s,a) + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}}Q(s',a') - Q(s,a)\right) $  
  and the following equation:  
  $ \theta \leftarrow \theta + \alpha\left(r+\gamma \max_{a' \in \mathcal{A}} Q_{\theta}(s', a') - Q_{\theta}(s, a)\right) \nabla_{\theta}Q_{\theta}(s, a) $   
  are exactly the same when this form of linear approximation is used.

#### STEP 1.   
The update rule for $\theta$ we got in 5-(a)-2 is the following.  
$\theta \leftarrow \theta + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} (\theta^\top \delta(s', a')) - \theta^\top \delta(s, a)\right) \delta(s, a)$

#### case 1. $(s'',a'') \neq (s,a)$
In this update rule, for entry $(s'',a'') \neq (s,a)$, we have $[\delta(s, a)]_{(s'',a'')} = 0$   
and thus the update rule becomes $[\theta]_{(s'',a'')} \leftarrow [\theta]_{(s'',a'')}$

#### case 2. $(s'',a'') = (s,a)$
In this update rule, for entry $(s'',a'') = (s,a)$, we have $[\delta(s, a)]_{(s'',a'')} = 1$   
and thus the update rule becomes $[\theta]_{(s'',a'')} \leftarrow [\theta]_{(s'',a'')} + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} (\theta^\top \delta(s', a')) - \theta^\top \delta(s, a)\right)$  

or equivalently,
$[\theta]_{(s,a)} \leftarrow [\theta]_{(s,a)} + \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} (\theta^\top \delta(s', a')) - \theta^\top \delta(s, a)\right)$

#### STEP 2. Review the LINEAR APPROXIMATION of Q.  
###Disclaimer : Here we use (s,a) to represent the ($s*|S|+a$)-entry for simplicity.

Let's review the linear approximation $Q_{\theta} (s,a) = \theta^T\delta(s,a) = [0,0,...,0,0,[\theta]_{(s,a)},0,0,...0]$, which is a |S||A|-dimension row vector having all entries zero except for the (s,a)-entry.
i.e.

$[Q_{\theta} (s,a)]_{(s'',a'')} = 0$  if $(s'',a'') \neq  (s,a)$  
$[Q_{\theta} (s,a)]_{(s'',a'')} = [\theta]_{(s,a)}$ if $(s'',a'') =  (s,a)$

### STEP 3. SUMMARY
Therefore, the last formula in STEP 1 can be further reduced to

$[Q_{\theta} (s,a)]_{(s,a)} \leftarrow [Q_{\theta} (s,a)]_{(s,a)}+ \alpha\left(r + \gamma \max_{a' \in \mathcal{A}} (\theta^\top \delta(s', a')) - \theta^\top \delta(s, a)\right)$   

which is exactly the same as the update rule under the tabular setting when this form of linear approximation is used.